{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import simulator\n",
    "import csv\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_dict = {\"0/0\" : 0,\n",
    "                \"0/1\" : 1,\n",
    "                \"1/0\" : 1,\n",
    "                \"1/1\" : 2,\n",
    "                \"0/.\" : 0,\n",
    "                \"./0\" : 0,\n",
    "                \"1/.\" : 2,\n",
    "                \"./1\" : 2,\n",
    "                \"./.\" : -1}\n",
    "\n",
    "def compare_VCF_sites(real_f, inferred_f, n_sites):\n",
    "    real_f.seek(0)\n",
    "    inferred_f.seek(0)\n",
    "    true_reader = list(csv.reader(real_f, delimiter='\\t'))\n",
    "    call_reader = list(csv.reader(inferred_f, delimiter='\\t'))\n",
    "    TP = FN = TN = FP = 0\n",
    "    #Sites only\n",
    "    for true_line in true_reader:\n",
    "        found = False\n",
    "        #True Positive\n",
    "        for call_line in call_reader:\n",
    "            if call_line[0] == true_line[0] and call_line[1] == true_line[1]:\n",
    "                TP += 1\n",
    "                found = True\n",
    "                break\n",
    "        #False Negative\n",
    "        if not found:\n",
    "            FN += 1\n",
    "    for call_line in call_reader:\n",
    "        found = False\n",
    "        #True Positive (again, not counting)\n",
    "        for true_line in true_reader:\n",
    "            if call_line[0] == true_line[0] and call_line[1] == call_line[1]:\n",
    "                found = True\n",
    "                break\n",
    "        #False Positive\n",
    "        if not found:\n",
    "            FP += 1\n",
    "    #True negatives\n",
    "    TN = n_sites - TP - FN - FP\n",
    "    precision = TP/(TP + FP)\n",
    "    recall    = TP/(TP + FN)\n",
    "    F1        = (2*TP)/(2*TP + FP + FN)\n",
    "    return (precision, recall, F1)\n",
    "    #return (TP,FN,FP,TN)\n",
    "    \n",
    "\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_VCF_cells(real_f, inferred_f, n_sites):\n",
    "    real_f.seek(0)\n",
    "    inferred_f.seek(0)\n",
    "    true_reader = list(csv.reader(real_f, delimiter='\\t'))\n",
    "    call_reader = list(csv.reader(inferred_f, delimiter='\\t'))\n",
    "    #Cells\n",
    "    TP = FN = TN = FP = 0\n",
    "    for i, field in enumerate(true_reader[0]):\n",
    "        if field in genotype_dict.keys():\n",
    "            first_cell_col_t = i\n",
    "            break\n",
    "    for i, field in enumerate(call_reader[0]):\n",
    "        if field in genotype_dict.keys():\n",
    "            first_cell_col_c = i\n",
    "            break\n",
    "    m = len(true_reader[0]) - first_cell_col_t\n",
    "    for true_line in true_reader:\n",
    "        found = False\n",
    "        for call_line in call_reader:\n",
    "            if call_line[0] == true_line[0] and call_line[1] == true_line[1]:\n",
    "                #Correct site call found\n",
    "                found = True\n",
    "                for i in range(m):\n",
    "                    if genotype_dict[true_line[first_cell_col_t + i]] > 0:\n",
    "                        #Cell is really mutant\n",
    "                        if genotype_dict[call_line[first_cell_col_c + i]] > 0:\n",
    "                            #And called correctly\n",
    "                            TP += 1\n",
    "                        else:\n",
    "                            #Real mutant not called\n",
    "                            FN += 1\n",
    "                    else:\n",
    "                        #Cell is welltype\n",
    "                        if genotype_dict[call_line[first_cell_col_c + i]] > 0:\n",
    "                            #But called as mutant\n",
    "                            FP += 1\n",
    "                        else:\n",
    "                            #Real wt called as wt\n",
    "                            TN += 1\n",
    "                break\n",
    "        if not found:\n",
    "            # Site mutant but not called\n",
    "            for i in range(m):\n",
    "                if genotype_dict[true_line[first_cell_col_t + i]] > 0:\n",
    "                    #Cell actually mutant\n",
    "                    FN += 1\n",
    "                else:\n",
    "                    TN += 1\n",
    "    for call_line in call_reader:\n",
    "        found = False\n",
    "        for true_line in true_reader:\n",
    "            if call_line[0] == true_line[0] and call_line[1] == call_line[1]:\n",
    "                #Already dealt with correctly called sites above\n",
    "                found = True\n",
    "                break\n",
    "        #Site called with no real mutation\n",
    "        if not found:\n",
    "            for i in range(m):\n",
    "                if genotype_dict[call_line[first_cell_col_c + i]] > 0:\n",
    "                    #Cell called mutant\n",
    "                    FP += 1\n",
    "    TN = m * n_sites - TP - FP - FN\n",
    "    precision = TP/(TP + FP)\n",
    "    recall    = TP/(TP + FN)\n",
    "    F1        = (2*TP)/(2*TP + FP + FN)\n",
    "    return (precision, recall, F1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.9269406392694064, 0.9982619364073203, 0.9248291571753986)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "real_vcf = open(\"temp_r.vcf\", \"r\")\n",
    "call_vcf = open(\"temp_c.vcf\", \"r\")\n",
    "compare_VCF_cells(real_vcf, call_vcf, 2000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_params(m_cells, iters, params):\n",
    "    site_results = []\n",
    "    cell_results = []\n",
    "    for i in range(iters):\n",
    "        T = simulator.Phylogeny()\n",
    "        T.evolve(n_generations=1000, germline=True)\n",
    "        T.evolve(n_cells=5, germline=False)\n",
    "        m = len(T.active_nodes)\n",
    "        T.prepare()\n",
    "        vcf_f = open(\"temp_r.vcf\", \"w+\")\n",
    "        T.write_vcf(vcf_f)\n",
    "        vcf_f.close()\n",
    "        pfile = open(\"temp.pileup\", \"w+\")\n",
    "        T.write_pileup(pfile)\n",
    "        pfile.close()\n",
    "        args = [\"../SCarborSNV\", \"-m\", str(m),\"-p\", \"temp.pileup\", \"-o\" \"temp_c.vcf\"]\n",
    "        for name, val in params.items():\n",
    "            args.append(\"--{}={}\".format(name, val))\n",
    "        subprocess.run(args)\n",
    "        real_vcf = open(\"temp_r.vcf\", \"r\")\n",
    "        call_vcf = open(\"temp_c.vcf\", \"r\")\n",
    "        site_results.append(compare_VCF_sites(real_vcf, call_vcf, 2000))\n",
    "        cell_results.append(compare_VCF_cells(real_vcf, call_vcf, 2000))\n",
    "        real_vcf.close()\n",
    "        call_vcf.close()\n",
    "    os.remove(\"temp.pileup\")\n",
    "    os.remove(\"temp_r.vcf\")\n",
    "    os.remove(\"temp_c.vcf\")\n",
    "    return (site_results, cell_results)\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([(0.31645569620253167, 0.8333333333333334, 0.45871559633027525),\n",
       "  (0.5970149253731343, 0.6557377049180327, 0.625),\n",
       "  (0.6, 0.717391304347826, 0.6534653465346535)],\n",
       " [(0.29923273657289, 1.0, 0.46062992125984253),\n",
       "  (0.49491525423728816, 0.9299363057324841, 0.6460176991150443),\n",
       "  (0.5106382978723404, 0.96, 0.6666666666666666)])"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_params(6, 3, {\"posterior-threshold\": 0})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/oldham/.local/lib/python3.5/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-121-0483e7236dca>\", line 12, in <module>\n",
      "    site_res, cell_res = test_params(4, 2, prms) #10, 10\n",
      "  File \"<ipython-input-94-d88223bb6aa3>\", line 21, in test_params\n",
      "    call_vcf = open(\"temp_c.vcf\", \"r\")\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'temp_c.vcf'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/oldham/.local/lib/python3.5/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'FileNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/oldham/.local/lib/python3.5/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/oldham/.local/lib/python3.5/site-packages/IPython/core/ultratb.py\", line 319, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/oldham/.local/lib/python3.5/site-packages/IPython/core/ultratb.py\", line 353, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/usr/lib/python3.5/inspect.py\", line 1453, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/usr/lib/python3.5/inspect.py\", line 1410, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/usr/lib/python3.5/inspect.py\", line 672, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/usr/lib/python3.5/inspect.py\", line 701, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/usr/lib/python3.5/inspect.py\", line 685, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/usr/lib/python3.5/posixpath.py\", line 361, in abspath\n",
      "    cwd = os.getcwd()\n",
      "FileNotFoundError: [Errno 2] No such file or directory\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'temp_c.vcf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"
     ]
    }
   ],
   "source": [
    "#Finding best thresholds with everything else default, 10 cells\n",
    "posterior_threshs = [0.01,0.02] #[0, 1.0e-9, 1.0e-6, 1.0e-3, 5.0e-3, 0.01, 0.02, 0.05, 0.1, 0.15, 0.2, 0.3, 0.5, 1]\n",
    "candidate_threshs = [0.5, 0.6] #[0.3, 0.4, 0.5, 0.6, 0.7, 0.9]\n",
    "site_data = pd.DataFrame(columns=[\"cand\",\"post\",\"precision\",\"recall\",\"F1\"])\n",
    "cell_data = pd.DataFrame(columns=[\"cand\",\"post\",\"precision\",\"recall\",\"F1\"])\n",
    "i = 0\n",
    "for t1 in candidate_threshs:\n",
    "    for t2 in posterior_threshs:\n",
    "        prms = {}\n",
    "        prms[\"candidate-threshold\"] = t1\n",
    "        prms[\"posterior-threshold\"] = t2\n",
    "        site_res, cell_res = test_params(4, 2, prms) #10, 10\n",
    "        for j in range(2): #10\n",
    "            site_data.loc[i] = [t1, t2, *site_res[j]]\n",
    "            cell_data.loc[i] = [t1, t2, *cell_res[j]]\n",
    "            i += 1\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_data.to_csv(\"cell_data.csv\", index=False)\n",
    "site_data.to_csv(\"site_data.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>post</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0.000000e+00</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.659233</td>\n",
       "      <td>0.960835</td>\n",
       "      <td>0.748327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.000000e-02</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.532046</td>\n",
       "      <td>0.981102</td>\n",
       "      <td>0.622124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2.000000e-01</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.524656</td>\n",
       "      <td>0.984976</td>\n",
       "      <td>0.614063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5.000000e-02</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.521941</td>\n",
       "      <td>0.982797</td>\n",
       "      <td>0.614587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.000000e-06</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.504793</td>\n",
       "      <td>0.986944</td>\n",
       "      <td>0.606669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.000000e-01</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.497375</td>\n",
       "      <td>0.985726</td>\n",
       "      <td>0.593066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.500000e-01</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.489890</td>\n",
       "      <td>0.986256</td>\n",
       "      <td>0.584594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2.000000e-02</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.479201</td>\n",
       "      <td>0.984458</td>\n",
       "      <td>0.564054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.000000e-09</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.462629</td>\n",
       "      <td>0.977730</td>\n",
       "      <td>0.570449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.000000e+00</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.445642</td>\n",
       "      <td>0.987130</td>\n",
       "      <td>0.561443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.000000e-01</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.434687</td>\n",
       "      <td>0.982136</td>\n",
       "      <td>0.541366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5.000000e-03</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.432818</td>\n",
       "      <td>0.983832</td>\n",
       "      <td>0.541653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.000000e-03</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.430616</td>\n",
       "      <td>0.986817</td>\n",
       "      <td>0.530932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5.000000e-01</th>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.427472</td>\n",
       "      <td>0.983092</td>\n",
       "      <td>0.523585</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  post  precision    recall        F1\n",
       "cand                                                 \n",
       "0.000000e+00  0.566667   0.659233  0.960835  0.748327\n",
       "1.000000e-02  0.566667   0.532046  0.981102  0.622124\n",
       "2.000000e-01  0.566667   0.524656  0.984976  0.614063\n",
       "5.000000e-02  0.566667   0.521941  0.982797  0.614587\n",
       "1.000000e-06  0.566667   0.504793  0.986944  0.606669\n",
       "1.000000e-01  0.566667   0.497375  0.985726  0.593066\n",
       "1.500000e-01  0.566667   0.489890  0.986256  0.584594\n",
       "2.000000e-02  0.566667   0.479201  0.984458  0.564054\n",
       "1.000000e-09  0.566667   0.462629  0.977730  0.570449\n",
       "1.000000e+00  0.566667   0.445642  0.987130  0.561443\n",
       "3.000000e-01  0.566667   0.434687  0.982136  0.541366\n",
       "5.000000e-03  0.566667   0.432818  0.983832  0.541653\n",
       "1.000000e-03  0.566667   0.430616  0.986817  0.530932\n",
       "5.000000e-01  0.566667   0.427472  0.983092  0.523585"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#After running full grid search on other machine\n",
    "cell_results = pd.read_csv(\"threshold_cell_data.csv\")\n",
    "cell_results.groupby(['cand']).mean().sort_values(\"precision\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "#argnames = [\"lambda\", \"mu\", \"p-haploid\", \"p-clonal\",\"pileup-file\", \"vcf-file\",\"n-cells\",\"temp-file\",\"amp-err\",\"p-ado\",\"candidate-threshold\",\"posterior-threshold\"] "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
